xdsl/backend/riscv/lowering/convert_memref_to_riscv.py

```diff
@@ -0,0 +1,135 @@
+from collections.abc import Sequence
+from typing import Any, cast
+
+from xdsl.backend.riscv.lowering.utils import (
+    cast_operands_to_regs,
+    register_type_for_type,
+)
+from xdsl.dialects import memref, riscv
+from xdsl.dialects.builtin import (
+    ModuleOp,
+    UnrealizedConversionCastOp,
+)
+from xdsl.ir import MLContext, SSAValue
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.exceptions import DiagnosticException
+
+
+class ConvertMemrefAllocOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.Alloc, rewriter: PatternRewriter) -> None:
+        raise DiagnosticException("Lowering memref.alloc not implemented yet")
+
+
+class ConvertMemrefDeallocOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.Dealloc, rewriter: PatternRewriter) -> None:
+        raise DiagnosticException("Lowering memref.dealloc not implemented yet")
+
+
+def insert_shape_ops(
+    mem: SSAValue,
+    indices: Sequence[SSAValue],
+    shape: Sequence[int],
+    rewriter: PatternRewriter,
+) -> SSAValue:
+    """
+    Returns ssa value representing pointer into the memref at given indices.
+    """
+    assert len(shape) == len(indices)
+
+    match indices:
+        case [idx1]:
+            rewriter.insert_op_before_matched_op(
+                [
+                    ptr := riscv.AddOp(mem, idx1),
+                ]
+            )
+        case [idx1, idx2]:
+            rewriter.insert_op_before_matched_op(
+                [
+                    cols := riscv.LiOp(shape[1]),
+                    row_offset := riscv.MulOp(cols, idx1),
+                    offset := riscv.AddOp(row_offset, idx2),
+                    offset_bytes := riscv.SlliOp(
+                        offset, 2, comment="mutiply by elm size"
```

Comment:
I feel like I should know this, but we are always multiplying by 4, right?
It's not dependent on the type, right?

Reply:
hmm 4 because 4 bytes in a word on 32-bit systems but you're right that it's a tricky thing, I'll add a comment

- [ ] add a comment